# Overview of CD4+ Cells .h5ad files
**Dataset:** GWT Perturb-seq Analysis — CD4+ T Cells

More info at https://virtualcellmodels.cziscience.com/dataset/genome-scale-tcell-perturb-seq

All the relevant files are located in `.../Ilaria/Data/`.

- Filename: `<Donor ID>_<Cell Condition>.assigned_guide.h5ad` (e.g., `D1_Rest.assigned_guide.h5ad`)
- Donor IDs: D1, D2, D3, D4
- Cell Conditions: Rest, Stim8hr, Stim48hr

### Workspace Setup

In [1]:
# Check working directory
import os
print(f"Current directory: {os.getcwd()}")

Current directory: /mnt/R0/Projects/POIAZ/Ilaria/Scripts


In [3]:
# Import packages
import numpy as np
import anndata
import pandas as pd
import mudata as md
import scanpy as sc

In [3]:
# Setting up filenames
donors = ['D1', 'D2', 'D3', 'D4']
cell_conditions = ['Rest', 'Stim8hr', 'Stim48hr']
filenames = list()
for d in donors:
    for cc in cell_conditions:
        filenames.append(d + '_' + cc + '.assigned_guide.h5ad')
        print(d + '_' + cc + '.assigned_guide.h5ad')

D1_Rest.assigned_guide.h5ad
D1_Stim8hr.assigned_guide.h5ad
D1_Stim48hr.assigned_guide.h5ad
D2_Rest.assigned_guide.h5ad
D2_Stim8hr.assigned_guide.h5ad
D2_Stim48hr.assigned_guide.h5ad
D3_Rest.assigned_guide.h5ad
D3_Stim8hr.assigned_guide.h5ad
D3_Stim48hr.assigned_guide.h5ad
D4_Rest.assigned_guide.h5ad
D4_Stim8hr.assigned_guide.h5ad
D4_Stim48hr.assigned_guide.h5ad


### Loading Summary Files

Upload and print overview of DE_stats and pseudobulk_merged.

In [4]:
# Load GWCD4i.DE_stats file with AnnData
DE_stats = anndata.read_h5ad('../Data/GWCD4i.DE_stats.h5ad', backed="r")
print(f"DE_stats loaded: {DE_stats.shape}")
print(f"Observations (perturbation-condition pairs): {DE_stats.n_obs}")
print(f"Variables (genes): {DE_stats.n_vars}")

# Print list of .obs
print("\n--- .obs columns ---")
print(DE_stats.obs.columns.tolist())

# Print list of .layers
print("\n--- Available layers ---")
print(list(DE_stats.layers.keys()))

# Print count per condition
print("\n--- Culture conditions ---")
print(DE_stats.obs['culture_condition'].value_counts())

DE_stats loaded: (33983, 10282)
Observations (perturbation-condition pairs): 33983
Variables (genes): 10282

--- .obs columns ---
['target_contrast_gene_name', 'culture_condition', 'target_contrast', 'chunk', 'n_cells_target', 'n_up_genes', 'n_down_genes', 'n_total_de_genes', 'ontarget_effect_size', 'ontarget_significant', 'target_baseMean', 'offtarget_flag', 'n_total_genes_category', 'ontarget_effect_category', 'n_downstream']

--- Available layers ---
['adj_p_value', 'baseMean', 'lfcSE', 'log_fc', 'p_value', 'zscore']

--- Culture conditions ---
culture_condition
Stim8hr     11415
Rest        11287
Stim48hr    11281
Name: count, dtype: int64


This AnnData object contains genome-wide differential expression results from a perturb-seq experiment in CD4+ T cells. Each observation represents a single perturbation (perturbed gene) tested in a specific culture condition (n_obs = 33,983). Each variable is a measured gene in the transcriptome (n_vars = 10,282).

Annotations for each perturbation-condition pair (`.obs`):
- `target_contrast_gene_name`: Name of the perturbed gene
- `culture_condition`: culture condition (Rest, Stim8hr, Stim48hr)
- `target_contrast`: Unique identifier for the perturbed gene
- `chunk`: differential expression processing group identifier
- `n_cells_target`: Number of cells with targeting guide for the perturbed gene
- `n_up_genes`: Count of significantly upregulated genes (10% FDR)
- `n_down_genes`: Count of significantly downregulated genes (10% FDR)
- `n_total_de_genes`: Total number of significantly differentially expressed genes (10% FDR)
- `ontarget_effect_size`: Effect size of the perturbation on its intended target gene
- `ontarget_significant`: Boolean indicating whether on-target knockdown was significant (10% FDR)
- `target_baseMean`: Mean baseline expression of the target gene
- `offtarget_flag`: Flag indicating potential off-target effects (TSS within 10 kb with significant down-regulation)
- `n_total_genes_category`: Category based on number of trans-effects
- `ontarget_effect_category`: Category based on on-target / off-target effects
- `n_downstream`: Number of genes significantly affected by this perturbation, excluding on-target effect (incoming trans-effects)

Summary statistics for measured genes across conditions (`.varm`):
- `measured_genes_stats_Stim8hr`: Gene-level statistics for 8-hour stimulation condition
- `measured_genes_stats_Stim48hr`: Gene-level statistics for 48-hour stimulation condition
- `measured_genes_stats_Rest`: Gene-level statistics for resting/unstimulated condition

Differential expression statistics for each perturbation-gene pair from DESeq2 (`.layers`):
- `adj_p_value`: FDR-adjusted p-values,
- `baseMean`: Mean normalized expression of the gene across cells,
- `lfcSE`: Standard error of log fold change,
- `log_fc`: Log2 fold change,
- `p_value`: Raw p-values from differential expression testing,
- `zscore`: Z-scores for differential expression (logFC / lfcSE).


In [5]:
# Load GWCD4i.pseudobulk_merged file with AnnData
pseudobulk = anndata.read_h5ad('../Data/GWCD4i.pseudobulk_merged.h5ad', backed="r")
print(f"Pseudobulk loaded: {pseudobulk.shape}")
print(f"Observations (pseudobulk samples): {pseudobulk.n_obs}")
print(f"Variables (genes): {pseudobulk.n_vars}")

# Print list of .obs
print("\n--- .obs columns ---")
print(pseudobulk.obs.columns.tolist())

# Print list of .layers
print("\n--- Available layers ---")
print(list(pseudobulk.layers.keys()))

# Print count per condition
print("\n--- Culture conditions ---")
print(pseudobulk.obs['culture_condition'].value_counts())

Pseudobulk loaded: (278684, 18129)
Observations (pseudobulk samples): 278684
Variables (genes): 18129

--- .obs columns ---
['10xrun_id', 'donor_id', 'culture_condition', 'guide_id', 'perturbed_gene_name', 'perturbed_gene_id', 'guide_type', 'n_cells', 'total_counts', 'keep_min_cells', 'keep_effective_guides', 'keep_total_counts', 'keep_for_DE', 'keep_test_genes', 'log10_n_cells']

--- Available layers ---
[]

--- Culture conditions ---
culture_condition
Stim8hr     94149
Stim48hr    92693
Rest        91842
Name: count, dtype: int64


This AnnData object contains pseudobulk expression profiles. Each observation represents a pseudobulk (aggregated by guide, donor and culture condition). Each variable is a measured gene in the transcriptome (n_vars = 18,129).

Annotations for each pseudobulk sample (`.obs`):
- `10xrun_id`: processing batch identifier (R1 or R2)
- `donor_id`: Donor identifier
- `culture_condition`: Culture condition (Rest, Stim8hr, Stim48hr)
- `guide_id`: Unique guide identifier
- `perturbed_gene_name`: Name of the gene perturbed by the guide (note that the annotated gene in the guide identifier doesn't always match because we did some post-hoc curation of the target gene)
- `perturbed_gene_id`: Ensembl gene ID of the perturbed gene
- `guide_type`: Type of guide (e.g., targeting, non-targeting)
- `n_cells`: Number of cells aggregated in this pseudobulk sample
- `total_counts`: Total UMI counts across all cells in this pseudobulk
    log10_n_cells`: Log10-transformed number of cells
- `keep_min_cells`: Boolean flag indicating sample passes minimum cell count threshold to be used for DE analysis
- `keep_effective_guides`: Boolean flag indicating guide was considered effective (t-test significant) to be used for DE analysis
- `keep_total_counts`: Boolean flag indicating sample passes total counts threshold to be used for DE analysis
- `keep_for_DE`: Boolean flag indicating sample is suitable for differential expression analysis
- `keep_test_genes`: Boolean flag indicating whether the perturbed gene passes criteria for differential expression analysis

Expression Matrix (.X):
- Sum of UMI counts across cells for each gene in each pseudobulk sample


In [6]:
# Load D1_Rest.assigned_guide.h5ad file with AnnData
D1_Rest = anndata.read_h5ad('../Data/D1_Rest.assigned_guide.h5ad', backed='r')
print(f"Pseudobulk loaded: {D1_Rest.shape}")
print(f"Observations (pseudobulk samples): {D1_Rest.n_obs}")
print(f"Variables (genes): {D1_Rest.n_vars}")
print(f"Layers: {list(D1_Rest.layers.keys())}")
print(f"Observations: {list(D1_Rest.obs}")

# Print list of .obs
print("\n--- .obs columns ---")
print(pseudobulk.obs.columns.tolist())

# Print list of .layers
print("\n--- Available layers ---")
print(list(pseudobulk.layers.keys()))

# Print count per condition
print("\n--- Culture conditions ---")
print(pseudobulk.obs['culture_condition'].value_counts())

Pseudobulk loaded: (3074496, 18130)
Observations (pseudobulk samples): 3074496
Variables (genes): 18130


Each AnnData object contains cell expression profiles for cells from one donor (D1, D2, D3, D4) and culture condition (Rest, Stim8hr, Stim48hr). Cells from different 10X lanes are concatenated. Each observation represents a cell. Each variable is a measured gene in the transcriptome.

Annotations for each single cell (`.obs`):
- `lane_id`: 10X lane identifier (corresponds to one cellranger output)
- `n_genes_by_counts`: Number of genes with non-zero counts detected in the cell
- `total_counts`: Total UMI counts in the cell
- `pct_counts_mt`: Percentage of counts mapping to mitochondrial genes
- `top_guide_UMI_counts`: UMI counts for the most abundant guide RNA in the cell
- `guide_id`: Unique identifier for the guide RNA detected in the cell (if more than one guide was detected, we annotate as "multi-guide")
- `perturbed_gene_name`: Name of the gene perturbed by the detected guide (before target curation)
- `perturbed_gene_id`: Ensembl gene ID of the perturbed gene (before target curation)
- `guide_type`: Type of guide (e.g., targeting, non-targeting)
- `PuroR`: Puromycin resistance marker expression level
- `guide_group`: Group classification for the guide
- `low_quality`: Boolean flag indicating low-quality cells to be filtered

In [ ]:
# Quick structure overview
x = D1_Rest # choose object to view

print("=== AnnData structure ===")
print("\n--- .obs columns ---")
print(x.obs.columns.tolist())
print("\n--- culture_condition values ---")
print(x.obs['culture_condition'].value_counts())
print("\n--- layers available ---")
print(list(x.layers.keys()))

## Quick Tests
Code used to identify non-targeting control group inside the data.

In [ ]:
# Search for NTC in pseudobulk data
print("=== Searching for NTC in Pseudobulk ===")

# Check guide_type for non-targeting
if 'guide_type' in pseudobulk.obs.columns:
    ntc_pseudobulk = pseudobulk.obs['guide_type'].str.contains('non-targeting', na=False)
    print(f"\nRows with 'non-targeting' in guide_type: {ntc_pseudobulk.sum()}")
    
    if ntc_pseudobulk.sum() > 0:
        print("\nNTC pseudobulk samples by condition:")
        ntc_samples = pseudobulk.obs[ntc_pseudobulk]
        if 'culture_condition' in ntc_samples.columns:
            print(ntc_samples['culture_condition'].value_counts())
        print(f"\nTotal NTC pseudobulk samples: {len(ntc_samples)}")
        print("\nSample of NTC entries:")
        print(ntc_samples.head())